### PR TITLE
new API

### PR DESCRIPTION
**Pull request by jcarres-mdsol**
The API is:
Farscape.invoke(url).invoke('create', {title: new_title}).properties
Farscape.invoke(url).invoke('self').properties

@fosdev  please review

All Green  Ready to merge

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -15,8 +15,8 @@ a response with a supported Hypermedia media-type and a root that lists availabl
 
 ### A Hypermedia API
 For a interacting with an API (or individual service that supports a list of resources at its root), you enter the
-API and follow your nose using the `enter` method on the agent. This method returns a [Farscape::Representor]()
```
Comment (jcarres-mdsol) [inline README.md:18]: Ignore changes in this file for now.

File: lib/farscape/representor.rb [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+      if link.nil?
+        raise UnknownTransition, "There was no link found for #{link_name} in #{@representor}"
+      end
+      Transition.new(link).invoke(data_as_a_hash)
+    end
+
+    # We are decorating Representor::Transition with Farscape::Transition so they can be invoked
+    # @return [Array] of Farscape::Transition
+    def transitions
+      super().map {|transition_data| Transition.new(transition_data)}
```
Comment (fosdev) [inline lib/farscape/representor.rb:35]: Don't need () on super. However, shouldn't you just use @representor.transitions.map instead? Also, I think this should memoize.

File: lib/farscape/representor.rb [hunk 3]
```diff
@@ -0,0 +1,69 @@
+module Farscape
+
+  # This is the main class the user of this library is gonna use to interact with the data
+  # The data itself is accessed via @representor. The other methods are added by farscape
+  # to it provides an easy to use API.
+  # study.invoke('lock') should give back a new representor that can be acted upon or read data from
+  class Representor < BasicObject
```
Comment (fosdev) [inline lib/farscape/representor.rb:7]: Why not just make this a decorator class the extends Representors::Representor vs. a delegator class, effectively? Ditto on Farscape Transition?

``` ruby
module Farscape
  class Representor < Representors::Representor
    def initialize(representor, result, &block)
      super(representor.to_representor_hash, &block)
      @result = result
   end

   # other methods
  end
end
```

Also, I wonder whether the result should go in the underlying representors object or here. I could go either way.

Comment (fosdev) [inline lib/farscape/representor.rb:7]: updated comment.

File: lib/farscape/representor.rb [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+      super().map {|transition_data| Transition.new(transition_data)}
+    end
+
+    # This provides:
+    # agent.headers
+    # agent.status
+    # agent.body  (raw response)
+    # agent.request
+    # agent.response
+    def agent
```
Comment (fosdev) [inline lib/farscape/representor.rb:44]: This should just be #last_result or something. I would expect agent to be an farscape instance.

File: spec/lib/farscape/transition_spec.rb [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+  describe Transition do
+    let(:empty_transition) { ::Representors::Transition.new({})}
+    let(:empty_farscape_transition) { Transition.new(empty_transition)}
+
+    describe '.new' do
+      it 'returns the correct class' do
+        expect(empty_farscape_transition.class).to eq(Farscape::Transition)
+      end
+
+      it '#instance_of? Transition is true' do
```
Comment (fosdev) [inline spec/lib/farscape/transition_spec.rb:13]: these should be examples under a describe '#instance_of?' block.

File: lib/farscape/representor.rb [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+      @result
+    end
+
+    # So we tell properly who we are and do not get users confused
+    def class
+      Representor
+    end
+
+    # So we tell properly who we are and do not get users confused
+    def instance_of?(class_name)
```
Comment (fosdev) [inline lib/farscape/representor.rb:54]: U probably won't need this if you implement according to my proposed decorator pattern above.

File: lib/farscape/representor.rb [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+      if class_name == Representor
+        true
+      else
+        false
+      end
+    end
+
+    # Any method called in this object, delegate it to the representor with the data
+    # so this works object.transitions
+    def method_missing(method, *args, &block)
```
Comment (fosdev) [inline lib/farscape/representor.rb:64]: Won't need this either.

File: lib/farscape/transition.rb [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+
+    # Main method for the users of this library to get the next piece of data from a service
+    # or to send data to a service
+    # @param [String] link_name. The link_name to follow
+    # @param [Hash] data_as_a_hash This is the data we want to provide if we are sending data
+    def invoke(data_as_a_hash = nil)
+      if data_as_a_hash
+        body = data_as_a_hash.to_json
+        # If the link does not provide a proper method, that is a bug in the reprensation we got
+        SimpleAgent.invoke(uri, {method: interface_method, body: body })
```
Comment (fosdev) [inline lib/farscape/transition.rb:26]: Have not accommodated passing parameters.

File: lib/farscape/unknown_transition.rb [hunk 9]
```diff
@@ -0,0 +1,5 @@
+module Farscape
```
Comment (fosdev) [inline lib/farscape/unknown_transition.rb:1]: Lets just have a farscape errors class and keep these all in one file including any base class extension they may use for request related errors. Thoughts?

**Comment by fosdev** [comment]
@JordiPolo Number of comments. However, per my swansong email to @mdsol/hypermedia-team, I really think we need to remove the #invoke method on the Farscape::Representor class and possibly add some transition related helpers. We can discuss that more in the thread of jump on an hangout sometime if need be.

Mark

**Comment by HonoreDB** [comment]
Sorry for letting this rot, closing.

**Pull request closed without merging**